### PR TITLE
[@ckeditor/ckeditor5-utils] Update to v25

### PR DESCRIPTION
**Pull request by fedemp**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://ckeditor.com/docs/ckeditor5/latest/api/utils.html
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by fedemp** [comment]
> @fedemp The CI build failed! Please [review the logs for more information](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/6879c9ccb22661894dc5135740c55c1f7dc66523/checks?check_suite_id=2078149361).
> 
> Once you've pushed the fixes, the build will automatically re-run. Thanks!

I run `test-all` locally several times and I couldn't find what is the right mapping. Could it be that the linter does not handle scoped and old packages?

`Error: Expected directory ckeditor__ckeditor5-utils/v10 to be the path mapping for @ckeditor/ckeditor5-utils`

File: types/ckeditor__ckeditor5-utils/src/emittermixin.d.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
+import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
```
Comment (peterblazejewicz) [inline types/ckeditor__ckeditor5-utils/src/emittermixin.d.ts:2]:
```suggestion
import EventInfo from "./eventinfo";
import { PriorityString } from "./priorities";
```
should those be relative?

File: types/ckeditor__ckeditor5-utils/src/fastdiff.d.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import { DeleteChange, InsertChange } from "@ckeditor/ckeditor5-utils/src/difftochanges";
```
Comment (peterblazejewicz) [inline types/ckeditor__ckeditor5-utils/src/fastdiff.d.ts:1]:
```suggestion
import { DeleteChange, InsertChange } from "./difftochanges";
```
shouldn't this one be relative as well?

File: types/ckeditor__ckeditor5-utils/src/focustracker.d.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { EmitterMixinDelegateChain, Emitter } from "@ckeditor/ckeditor5-utils/src/emittermixin";
+import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
+import { BindChain, Observable } from "@ckeditor/ckeditor5-utils/src/observablemixin";
+import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
```
Comment (peterblazejewicz) [inline types/ckeditor__ckeditor5-utils/src/focustracker.d.ts:4]:
```suggestion
import { EmitterMixinDelegateChain, Emitter } from "./emittermixin";
import EventInfo from "./eventinfo";
import { BindChain, Observable } from "./observablemixin";
import { PriorityString } from "./priorities";
```

File: types/ckeditor__ckeditor5-utils/src/keystrokehandler.d.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import { Emitter } from "@ckeditor/ckeditor5-utils/src/emittermixin";
+import { KeystrokeInfo } from "@ckeditor/ckeditor5-utils/src/keyboard";
+import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
```
Comment (peterblazejewicz) [inline types/ckeditor__ckeditor5-utils/src/keystrokehandler.d.ts:3]:
```suggestion
import { Emitter } from "./emittermixin";
import { KeystrokeInfo } from "./keyboard";
import { PriorityString } from "./priorities";
```

File: types/ckeditor__ckeditor5-utils/src/observablemixin.d.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
+import Emitter from '@ckeditor/ckeditor5-utils/src/emittermixin';
```
Comment (peterblazejewicz) [inline types/ckeditor__ckeditor5-utils/src/observablemixin.d.ts:1]:
```suggestion
import Emitter from './emittermixin';
```

**Comment by peterblazejewicz** [comment]
@fedem why using old `///` in `index.d.ts` instead copy/paste export/import from `src/index.js`? it should work a treat.
Why TS 3.6?

**Comment by fedemp** [comment]
> @FedeM why using old `///` in `index.d.ts` instead copy/paste exports/imports (with some adjustment ofc) from `src/index.js`? it should work a treat.
> 
> The expected usage is, I believe:
> 
> ```ts
> import * as utils from '@.....'
> ```

`tsc` actually created an `index.d.ts` with all the imports and exports, but since none of the examples in CKEditor website imports from `index.js` but rather from `src/*`, I wanted to somehow forbid consumers from using `index.d.ts`. But since it exists, I agree it should be added.
 
> Why TS 3.6?

When running `test-all` this was a lowest version that would not fail. Should I use latest TS version?

**Comment by peterblazejewicz** [comment]
> When running test-all this was a lowest version that would not fail. Should I use latest TS version?

Absolutely not, the concern here was to not cutoff users, who are (or could be) depending on older TS version, as the CKEditor is quite popular package out there.

**Comment by fedemp** [comment]
> > When running test-all this was a lowest version that would not fail. Should I use latest TS version?
> 
> Absolutely not, the concern here was to not cutoff users, who are (or could be) depending on older TS version, as the CKEditor is quite popular package out there.

So yes, `dslint` failed when it used TS 3.5. That's why I chose 3.6. :)

**Comment by peterblazejewicz** [comment]
Let's work one at time:
> Module '"@ckeditor/ckeditor5-utils/src/emittermixin"' has no exported member 'Emitter'. Did you mean to use 'import Emitter from "@ckeditor/ckeditor5-utils/src/emittermixin"' instead?

in tests, this is offending line:
```ts
import EmitterMixin, { Emitter } from '@ckeditor/ckeditor5-utils/src/emittermixin';
```

**Comment by peterblazejewicz** [comment]
Definition for `getPluralForm` is just as useful as using `any`,
can you consider to use (if I'm reading it ok):
```ts
export interface GetPluralForm {
    (n: number): boolean;
}
```
```ts
export function add(language: string, translations: Record<string, string>, getPluralForm?: GetPluralForm): void;
```

**Comment by fedemp** [comment]
> Definition for `getPluralForm` is just as useful as using `any`,
> can you consider to use (if I'm reading it ok):
> 
> ```ts
> export interface GetPluralForm {
>     (n: number): boolean;
> }
> ```
> 
> ```ts
> export function add(language: string, translations: Record<string, string>, getPluralForm?: GetPluralForm): void;
> ```

Done. :)

**Comment by fedemp** [comment]
> Let's work one at time:
> 
> > Module '"@ckeditor/ckeditor5-utils/src/emittermixin"' has no exported member 'Emitter'. Did you mean to use 'import Emitter from "@ckeditor/ckeditor5-utils/src/emittermixin"' instead?
> 
> in tests, this is offending line:
> 
> ```ts
> import EmitterMixin, { Emitter } from '@ckeditor/ckeditor5-utils/src/emittermixin';
> ```
> 
> those class + mixin imports are raising errors (iv'e fetched PR branch) in other places as well.
> For emitter and observable (concreate + mixins)

That's my bad. While typing the `engine` package I finally understood how ObservableMixin and EmitterMixin should work. I updated those definitions without updating the imports.

**Review by peterblazejewicz — COMMENTED** [review]
That looks really cool, 
what about one, final stroke to make it open for any TS version (3.2 or newer guaranted to work as tested by DT, but probably also in older versions). Up to you, I'll approve anyway:
```diff
commit b02cdf073c6ee86e518a679457cc03377c67656e
Author: Piotr B??a??ejewicz <peterblazejewicz@users.noreply.github.com>
Date:   Wed Feb 24 21:41:47 2021 +0100

    remove TS minimum version

diff --git a/types/ckeditor__ckeditor5-utils/index.d.ts b/types/ckeditor__ckeditor5-utils/index.d.ts
index ace5d8c375..9cd9f3991b 100644
--- a/types/ckeditor__ckeditor5-utils/index.d.ts
+++ b/types/ckeditor__ckeditor5-utils/index.d.ts
@@ -3,7 +3,6 @@
 // Definitions by: Federico Panico <https://github.com/fedemp>
 //                 denisname <https://github.com/denisname>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 3.6
 
 export { default as env } from "./src/env";
 export { default as diff } from "./src/diff";
diff --git a/types/ckeditor__ckeditor5-utils/src/collection.d.ts b/types/ckeditor__ckeditor5-utils/src/collection.d.ts
index 8ccb3ad0d4..2664eed184 100644
--- a/types/ckeditor__ckeditor5-utils/src/collection.d.ts
+++ b/types/ckeditor__ckeditor5-utils/src/collection.d.ts
@@ -57,17 +57,17 @@ export default class Collection<T> implements Iterable<T>, Emitter {
      * The number of items available in the collection.
      *
      */
-    get length(): number;
+    readonly length: number;
     /**
      * Returns the first item from the collection or null when collection is empty.
      *
      */
-    get first(): T | null;
+    readonly first: T | null;
     /**
      * Returns the last item from the collection or null when collection is empty.
      *
      */
-    get last(): T | null;
+    readonly last: T | null;
     /**
      * Adds an item into the collection.
      *
diff --git a/types/ckeditor__ckeditor5-utils/src/locale.d.ts b/types/ckeditor__ckeditor5-utils/src/locale.d.ts
index d85c53f329..d840e265a9 100644
--- a/types/ckeditor__ckeditor5-utils/src/locale.d.ts
+++ b/types/ckeditor__ckeditor5-utils/src/locale.d.ts
@@ -91,5 +91,5 @@ export default class Locale {
      * properties instead.
      *
      */
-    get language(): string;
+    readonly language: string;
 }
```

**Comment by fedemp** [comment]
Thanks. I have changed the getters you mentioned plus one more. Also restored the original minimum version of TS that was mentioned. Unfournately dtslint only tests from 3.4.

**Comment by peterblazejewicz** [comment]
> Thanks. I have changed the getters you mentioned plus one more. Also restored the original minimum version of TS that was mentioned. Unfournately dtslint only tests from 3.4.

yes, my intent in last comment was to show, one can remove the `// TypeScript` comment altogether. The DT tests everything with all LTS versions of TypeScript (current 3.4 as you said). You don't have to bind to specific version of language, unless your code does this or backward compatiblity support is rrequired. Once you're happy with the code, can you let folks on Discord channel, so they can review the CI problem and how to approach it, thx!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@fedemp thx!

**Comment by elibarzilay** [comment]
(Rebased to trigger a new build.)

**Review by peterblazejewicz — APPROVED** [review]

**Pull request merged**